### DISCUSSION TITLE
Why is Lua used for Xournal++'s plugin scripting and not Python?

### DISCUSSION THREAD

**Discussion by K4LCIFER** [xournalpp/xournalpp#2786] [Q&A]
This isn't a request as I am not educated enough on the matter, I'm just curious as to why this decision was made. Lua is decently old isn't super popular. Python is new(ish) and very readable for the average user. At first glance I would expect python to be the obvious choice. So why was Lua chosen as Xournal++'s plugin language?

**Comment by personalizedrefrigerator** [comment]
https://github.com/xournalpp/xournalpp/discussions/2559 seems related!

**Comment by Technius** [comment]
I'm not the person who created the plugin system, but here are points about each:

Lua:
* Small binary footprint (it's small enough to be commonly used in embedded systems). Easily bundled.
* Small language in terms of core concepts / abstractions. Easy to integrate with C/C++ code.
* Few batteries included. Either need to pull libraries down with LuaRocks or reinvent abstractions.
* [Unstable C ABI](http://lua-users.org/wiki/ApplicationBinaryInterface), but Lua 5.3 specifically (which we assume) is available in most distros.

Python:
* Large binary footprint, annoying to bundle.
* Inconsistent versioning across different distros. Python 2 vs Python 3 problem. Who knows if the `python` binary points to Python 2 or Python 3, or what major version of Python 3 is being used. Throw in questions of whether `pip` is even available or if a virtualenv is being used, and then you have a nightmare.
* Large language with many different core concepts / abstractions. Interaction with C/C++ code may be difficult.
* Many batteries included. Very powerful and flexible.
* [Unstable C ABI](https://docs.python.org/3/c-api/stable.html), which means you have to compile the Python interop on every different platform. For example, we must compile the interop separately for Python 3.6 and Python 3.7.

TL;DR Lua integration is a lot easier than Python integration in terms of both actual implementation code as well as packaging. You can do a lot more in Python but that comes at a cost.
